File: leetcode/21e.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode()
        cur = head
        l1_cur = l1
        l2_cur = l2
        while l1 or l2:
            if l1 and l2:
                if l1.val < l2.val:
                    l1_cur = l1
                    cur.next = l1
                    cur = l1
                    l1 = l1.next
                else:
                    l2_cur = l2
                    cur.next = l2
                    cur = l2
                    l2 = l2.next
            elif l1:
                l1_cur = l1
                cur.next = l1
                cur = l1
                l1 = l1.next
            elif l2:
                l2_cur = l2
                cur.next = l2
                cur = l2
                l2 = l2.next
                    
        return head.next

File: leetcode/21e.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        final ListNode sentinel = new ListNode();
        ListNode cur = sentinel;
        while (list1 != null || list2 != null) {
            if (list1 != null && list2 != null) { // both exist
                if (list1.val < list2.val) {
                    cur.next = list1;
                    list1 = list1.next;
                    cur = cur.next;
                } else {
                    cur.next = list2;
                    list2 = list2.next;
                    cur = cur.next;
                }
            } else if (list1 != null) {
                cur.next = list1;
                list1 = list1.next;
                cur = cur.next;
            } else if (list2 != null) {
                cur.next = list2;
                list2 = list2.next;
                cur = cur.next;
            }
        }
        return sentinel.next;
    }
}
